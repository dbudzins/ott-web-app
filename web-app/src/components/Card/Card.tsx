import React, { KeyboardEvent, memo } from 'react';
import classNames from 'classnames';
import { useTranslation } from 'react-i18next';

import Lock from '../../icons/Lock';

import styles from './Card.module.scss';

type CardProps = {
  onClick?: () => void;
  onHover?: () => void;
  title: string;
  posterSource?: string;
  tag: JSX.Element;
  progress?: number;
  posterAspect?: '1:1' | '2:1' | '2:3' | '4:3' | '5:3' | '16:9' | '9:16';
  featured?: boolean;
  disabled?: boolean;
  loading?: boolean;
  isCurrent?: boolean;
  isLocked?: boolean;
  currentLabel?: string;
  enableTitle?: boolean;
};

function Card({
  onClick,
  onHover,
  title,
  posterSource,
  tag,
  progress,
  posterAspect = '16:9',
  enableTitle = true,
  featured = false,
  disabled = false,
  loading = false,
  isCurrent = false,
  isLocked = true,
  currentLabel,
}: CardProps): JSX.Element {
  const { t } = useTranslation('common');
  const cardClassName = classNames(styles.card, {
    [styles.featured]: featured,
    [styles.disabled]: disabled,
  });
  const posterClassNames = classNames(styles.poster, styles[`aspect${posterAspect.replace(':', '')}`], {
    [styles.current]: isCurrent,
  });

  return (
    <div
      className={cardClassName}
      onClick={onClick}
      onMouseEnter={onHover}
      tabIndex={disabled ? -1 : 0}
      onKeyDown={(event: KeyboardEvent) => (event.key === 'Enter' || event.key === ' ') && !disabled && onClick && onClick()}
      role="button"
      aria-label={t('play_item', { title })}
    >
      <div className={posterClassNames} style={{ backgroundImage: posterSource ? `url(${posterSource})` : '' }}>
        {isCurrent && <div className={styles.currentLabel}>{currentLabel}</div>}
        {!loading && (
          <div className={styles.meta}>
            {featured && !disabled && enableTitle && <div className={classNames(styles.title, { [styles.loading]: loading })}>{title}</div>}
            <div className={styles.tags}>
              {isLocked && (
                <div className={classNames(styles.tag, styles.lock)} aria-label={t('card_lock')}>
                  <Lock />
                </div>
              )}
              {tag}
            </div>
          </div>
        )}
        {progress ? (
          <div className={styles.progressContainer}>
            <div className={styles.progressBar} style={{ width: `${Math.round(progress * 100)}%` }} />
          </div>
        ) : null}
      </div>
      {!featured && !disabled && enableTitle && (
        <div className={styles.titleContainer}>
          <div className={classNames(styles.title, { [styles.loading]: loading })}>{title}</div>
        </div>
      )}
    </div>
  );
}

export default memo(Card);
