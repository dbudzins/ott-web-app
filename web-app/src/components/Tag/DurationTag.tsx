import React from 'react';
import { useTranslation } from 'react-i18next';

import { logDev } from '#src/utils/common';

function DurationTag({ seconds }: { seconds: number }): JSX.Element | null {
  const { t } = useTranslation('common');
  logDev(t);
  if (!seconds) {
    return null;
  }

  const minutes = Math.ceil(seconds / 60);

  return <>{t('tag_duration_minutes', { minutes })}</>;
}

export default DurationTag;
