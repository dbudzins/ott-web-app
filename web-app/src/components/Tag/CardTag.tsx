import React from 'react';
import { useTranslation } from 'react-i18next';

import type { PlaylistItem } from '#types/playlist';
import Tag from '#src/components/Tag/Tag';
import DurationTag from '#src/components/Tag/DurationTag';

function CardTag({ item, includeTag = true }: { item: PlaylistItem; includeTag?: boolean }): JSX.Element | null {
  const { t } = useTranslation('common');

  const liveText = t('live');

  const getTagText = () => {
    if (item.seriesId) {
      return t('tag_series');
    }

    if (item.seasonNumber && item.seasonNumber !== '0' && item.episodeNumber) {
      return t('tag_season_episode', { seasonNumber: item.seasonNumber, episodeNumber: item.episodeNumber });
    }

    if (item.episodeNumber && item.episodeNumber !== '0') {
      return t('tag_episode', { episodeNumber: item.episodeNumber });
    }

    if (item.duration) {
      return <DurationTag seconds={item.duration} />;
    }

    if (item.duration === 0) {
      return liveText;
    }

    return null;
  };

  const tagText = getTagText();

  if (!tagText) {
    return null;
  }

  if (includeTag) {
    return <Tag isLive={tagText === liveText}>{tagText}</Tag>;
  }

  return <>{tagText}</>;
}

export default CardTag;
