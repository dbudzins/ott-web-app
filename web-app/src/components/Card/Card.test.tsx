import * as React from 'react';
import { render } from '@testing-library/react';
import type { PlaylistItem } from 'ott-common/types/playlist';

import Card from './Card';

import DurationTag from '#src/components/Tag/DurationTag';
import CardTag from '#src/components/Tag/CardTag';
import { renderWithi18n } from '#test/testUtils';

test('renders card with video title', async () => {
  const { container } = await renderWithi18n(<Card title="aa" tag={<DurationTag seconds={120} />} onClick={() => ''} />);
  expect(container).toMatchSnapshot();
});

test('Tag renders tag with Season and Episode', async () => {
  const { container } = await renderWithi18n(<CardTag item={{ seasonNumber: '2', episodeNumber: '5' } as PlaylistItem} />);
  expect(container).toMatchSnapshot();
});

test('Tag renders tag with Episode', async () => {
  const { container } = render(<CardTag item={{ episodeNumber: '5' } as PlaylistItem} />);
  expect(container).toMatchSnapshot();
});

test('Tag renders tag with Series', async () => {
  const { container } = render(<CardTag item={{ seriesId: '1235ABCD' } as PlaylistItem} />);
  expect(container).toMatchSnapshot();
});

test('Tag renders tag with duration', async () => {
  const { container } = render(<CardTag item={{ duration: 140 } as PlaylistItem} />);
  expect(container).toMatchSnapshot();
});

test('Tag renders tag with Live', async () => {
  const { container } = render(<CardTag item={{ duration: 0 } as PlaylistItem} />);
  expect(container).toMatchSnapshot();
});
