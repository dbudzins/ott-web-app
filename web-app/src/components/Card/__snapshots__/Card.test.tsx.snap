// Vitest Snapshot v1

exports[`Tag renders tag with Episode 1`] = `
<div>
  <div
    class="tag normal"
  >
    tag_episode
  </div>
</div>
`;

exports[`Tag renders tag with Live 1`] = `
<div>
  <div
    class="tag normal live"
  >
    live
  </div>
</div>
`;

exports[`Tag renders tag with Season and Episode 1`] = `
<div>
  <div
    class="tag normal"
  >
    tag_season_episode
  </div>
</div>
`;

exports[`Tag renders tag with Series 1`] = `
<div>
  <div
    class="tag normal"
  >
    tag_series
  </div>
</div>
`;

exports[`Tag renders tag with duration 1`] = `
<div>
  <div
    class="tag normal"
  >
    tag_duration_minutes
  </div>
</div>
`;

exports[`renders card with video title 1`] = `
<div>
  <div
    aria-label="play_item"
    class="card"
    role="button"
    tabindex="0"
  >
    <div
      class="poster aspect169"
    >
      <div
        class="meta"
      >
        <div
          class="tags"
        >
          <div
            aria-label="card_lock"
            class="tag lock"
          >
            <svg
              aria-hidden="true"
              class="icon"
              focusable="false"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12,17A2,2 0 0,0 14,15C14,13.89 13.1,13 12,13A2,2 0 0,0 10,15A2,2 0 0,0 12,17M18,8A2,2 0 0,1 20,10V20A2,2 0 0,1 18,22H6A2,2 0 0,1 4,20V10C4,8.89 4.9,8 6,8H7V6A5,5 0 0,1 12,1A5,5 0 0,1 17,6V8H18M12,3A3,3 0 0,0 9,6V8H15V6A3,3 0 0,0 12,3Z"
                fill="currentColor"
              />
            </svg>
          </div>
          tag_duration_minutes
        </div>
      </div>
    </div>
    <div
      class="titleContainer"
    >
      <div
        class="title"
      >
        aa
      </div>
    </div>
  </div>
</div>
`;
